pkg/controller/database/cosmosdb/managed.go

```diff
@@ -0,0 +1,165 @@
+	errNotNoSQLAccount    = "managed resource is not a Database Account"
+	errCreateNoSQLAccount = "cannot create Database Account"
+	errGetNoSQLAccount    = "cannot get Database Account"
+	errDeleteNoSQLAccount = "cannot delete Database Account"
+)
+
+// Setup adds a controller that reconciles NoSQLAccount.
+func Setup(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1beta1.CosmosDBAccountGroupKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1beta1.CosmosDBAccount{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1beta1.CosmosDBAccountGroupVersionKind),
+			managed.WithConnectionPublishers(),
+			managed.WithExternalConnecter(&connecter{kube: mgr.GetClient(), newClientFn: cosmosdb.NewDatabaseAccountClient}),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connecter struct {
+	kube        client.Client
+	newClientFn func(creds []byte) (*documentdb.DatabaseAccountsClient, error)
+}
+
+func (c *connecter) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	r, ok := mg.(*v1beta1.CosmosDBAccount)
+	if !ok {
+		return nil, errors.New(errNotNoSQLAccount)
+	}
+
+	p := &v1alpha3.Provider{}
+	n := meta.NamespacedNameOf(r.Spec.ProviderReference)
+	if err := c.kube.Get(ctx, n, p); err != nil {
+		return nil, errors.Wrapf(err, "cannot get provider %s", n)
+	}
+
+	if p.GetCredentialsSecretReference() == nil {
+		return nil, errors.New(errProviderSecretNil)
+	}
+
+	s := &corev1.Secret{}
+	n = types.NamespacedName{Namespace: p.Spec.CredentialsSecretRef.Namespace, Name: p.Spec.CredentialsSecretRef.Name}
+	if err := c.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrapf(err, "cannot get provider secret %s", n)
+	}
+
+	client, err := c.newClientFn(s.Data[p.Spec.CredentialsSecretRef.Key])
+	return &external{client: client}, errors.Wrap(err, "cannot create new Azure Database Account client")
+}
+
+// external is a createsyncdeleter using the Azure API.
+type external struct {
+	client *documentdb.DatabaseAccountsClient
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	r, ok := mg.(*v1beta1.CosmosDBAccount)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotNoSQLAccount)
+	}
+
+	res, err := e.client.CheckNameExists(ctx, meta.GetExternalName(r))
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errGetNoSQLAccount)
+	}
+
+	if res.Response.StatusCode == http.StatusNotFound {
+		return managed.ExternalObservation{ResourceExists: false}, nil
+	}
+
+	account, err := e.client.Get(ctx, r.Spec.ForProvider.ResourceGroupName, meta.GetExternalName(r))
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errGetNoSQLAccount)
+	}
+	v1beta1.UpdateCosmosDBAccountObservation(&r.Status, account)
```

Comment:
We typically include functions like this in the `client` pkg rather than the `api` pkg. However, this is more of a style preference than a hard and fast rule.